### PR TITLE
feat: read-only array

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Summary of changes

Detect when an array is read-only so that fewer constraints are generated

## Dependency

This PR relies on the dynamic array indexing PR #886 and must be merged after.

# Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [X] I have reviewed the changes on GitHub, line by line.
- [X] I have ensured all changes are covered in the description.
- [ ] This PR requires documentation updates when merged.

# Additional context

We should be able to use read-only arrays with barretenberg turbo-plonk implementation.

### CODE REVIEW

**Review by jfecher — COMMENTED** [review]
Looks like this PR contains changes from the dynamic indexing PR. Is it required?

**Comment by guipublic** [comment]
The PR can now be reviewed as #886 is merged

File: crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs [hunk 1]
```diff
@@ -25,21 +25,35 @@ use super::{
...
     index: Expression,
 }
 
 type MemAddress = u32;
 
+enum ArrayType {
+    Init(HashSet<MemAddress>, u32),
+    WriteOnly,
+    ReadOnly(Option<usize>),
+    ReadWrite(Option<usize>),
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs:40]: Can you add comments on each variant explaining what each field is for?

File: crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs [hunk 2]
```diff
@@ -55,12 +69,53 @@ impl ArrayHeap {
...
+            },
+            ArrayType::WriteOnly => {
+                if is_load {
+                    ArrayType::ReadWrite(None)
+                } else {
+                    ArrayType::WriteOnly
+                }
+            }
+            ArrayType::ReadOnly(last) => match (is_load, last) {
+                (true, Some(_)) => ArrayType::ReadWrite(None),
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs:104]: Why does a previously read only array become read/write if its 'last' is Some?

Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs:104]: Last indicates the first write to the array (after the initialisation phase), it seems strange to be read-only with a write, but if we have only writes at the end, we can discard them as nobody will read them.

File: crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs (hunk 1 above)
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs]: Done

Comment (jfecher) [reply crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs]: Can you explain what the fields are in the comments as well? E.g. what does the `Option<usize>` represent? And each field in Init?

File: crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs (hunk 2 above)
Comment (jfecher) [reply crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs:108]: > Last indicates the first write to the array

Why is it named `last` then?

Your explanation on the rest makes sense.

Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs:108]: It represent the position of the 'last writes'; it is the minimum position of the last writes

File: crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs (hunk 1 above)
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs]: Sure, I added more details.

File: crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs [hunk 3]
```diff
@@ -77,8 +139,21 @@ impl ArrayHeap {
...
+        match self.typ {
+            ArrayType::Init(_, _) | ArrayType::WriteOnly => len = 0,
+            ArrayType::ReadOnly(last) => {
+                len = last.unwrap_or(len);
+                read_write = false;
+            }
+            ArrayType::ReadWrite(last) => {
+                len = last.unwrap_or(len);
+            }
+        }
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs:156]:
```suggestion
let (len, read_write) = match self.typ {
    ArrayType::Init(_, _) | ArrayType::WriteOnly => (0, true),
    ArrayType::ReadOnly(last) => {
        (last.unwrap_or(len), false)
    }
    ArrayType::ReadWrite(last) => {
        (last.unwrap_or(len), true)
    }
};
```

File: crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs [hunk 4]
```diff
@@ -25,21 +25,42 @@ use super::{
...
     operation: Expression,
     value: Expression,
     index: Expression,
 }
 
 type MemAddress = u32;
 
+enum ArrayType {
+    /// Initialization phase: intializing the array with writes on the 0..array.len range
+    /// It contains the HashSet of the initialised indexes and the maximum of these indexes
```
Comment (TomAFrench) [inline crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs:38]:
```suggestion
    /// Initialization phase: initializing the array with writes on the 0..array.len range
    /// It contains the HashSet of the initialized indexes and the maximum of these indexes
```

File: crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs [hunk 5]
```diff
@@ -55,12 +76,53 @@ impl ArrayHeap {
...
+        self.typ = match &self.typ {
+            ArrayType::Init(init_idx, len) => match (is_load, index_const) {
+                (false, Some(idx)) => {
+                    let idx: MemAddress = idx.to_u128().try_into().unwrap();
+                    let mut init_idx2 = init_idx.clone();
+                    let mut len2 = *len;
+                    init_idx2.insert(idx);
+                    if idx + 1 > len2 {
+                        len2 = idx + 1;
+                    }
```
[the remark this answers is not in the document]
Comment (jfecher) [reply crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs:91]: Good catch, I think using `max` would be clearer here

**Review by jfecher — CHANGES_REQUESTED** [review]
Requesting fixes to the typo Tom caught and a small simplification

File: crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs]: Ok for the max

**Comment by guipublic** [comment]
> Thanks for keeping up with all the small suggestions. Apologies it took so many rounds

np, as long as you keep being reactive, it's fine!

**Pull request merged**